### DISCUSSION TITLE
Eager loading both sides of a self-referential foreign key

### DISCUSSION THREAD

**Discussion by ubernostrum** [sqlalchemy/sqlalchemy#10138] [Usage Questions] [answered]
I have a (SQLAlchemy 2.0) ORM model with a self-referential foreign key, declared similarly to [the "adjacency list" example in the documentation](https://docs.sqlalchemy.org/en/20/_modules/examples/adjacency_list/adjacency_list.html). I've declared both sides of the self-referential relationship with `lazy="selectin"`. However, only the "children" side actually eagerly loads; the "parent" side lazy loads unless I explicitly eager-load it at query time (i.e., via `.options(selectinload(...`).

[This gist](https://gist.github.com/ubernostrum/09f7523e4728839741e028652e307712) contains a standalone minimal example file which reproduces the behavior reliably for me; it crashes at the attempt to access the (not-yet-loaded) "parent" side of the relationship at line 81 of `eager_self_reference.py`. Also in that gist is console output showing Python version, installed packages, and the full traceback emitted from running the example.

Adjusting the `join_depth` of `children` does not appear to have an effect; I get the same behavior when setting it to higher values (tested with `join_depth=2` and `join_depth=3`).

I'm not sure what I'm missing here. Am I configuring/using this incorrectly, or is there something deeper going on?

**Comment by gordthompson** [comment]
Have you seen the ORM async example here?

https://docs.sqlalchemy.org/en/20/_modules/examples/asyncio/async_orm.html

Maybe you need to use `select(…).options(selectinload(…))`

**Reply by ubernostrum** [reply]
I have used the ORM async before, yes, and read quite a lot of the documentation.

And, yes, I can manually work around this with `options()`, as I already mentioned, but my belief is that should not be necessary -- the use of `lazy="selectin`" on the definition of the relationship (both directions) should cause it to be eagerly loaded without further manual intervention.

See also [this comment from @zzzeek  on my original Mastodon post about this](https://fosstodon.org/@zzzeek/110770994253124431).

**Comment by ubernostrum** [comment]
Incidentally, from some further tinkering, I can get it to work as expected by setting `join_depth` explicitly on *both* directions of the relationship. My understanding was that this shouldn't be necessary to declare on the "parent" side of the relationship, and in fact neither [the documentation on configuring self-referential eager loading](https://docs.sqlalchemy.org/en/20/orm/self_referential.html#configuring-self-referential-eager-loading) nor [the full example](https://docs.sqlalchemy.org/en/20/_modules/examples/adjacency_list/adjacency_list.html) do so -- both show setting `join_depth` only on the "children" side.

So that raises a couple of questions:

1. Is it intended behavior that `join_depth` must be set on both sides of the relation in order to configure eager loading for a self-referential relation?
2. If so, could the documentation/examples be adjusted to make that clear? As-is, copy/paste from the examples in the documentation will lead to a setup that fails to eager-load the "parent" side of the relation and may lead to crashes (as in my case).

**Comment by zzzeek** [comment, marked as answer]
OK so first off, here's "one weird trick" number one.    Why is there suddenly a huge need for eager loading to work 100% of the time in all cases?  because of asyncio and that there is no "lazy loading" pattern for asyncio.     SQLAlchemy 2.0.13 introduced a novel way to just work around this for a case like this, where you have this one line in your app that loads "node.parent" and depending on the query crashes because ".parent" wasn't loaded, and "we can't await an attribute".  Well now you **can** await an attribute by using [AsyncAttrs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs), so for this program it would be like this:

```py
class Node(sa_async.AsyncAttrs, Base):
    """
    A node in a tree.

    """
    # ...

# ... later ...

async def crash():
    session = get_sessionmaker()()
    two = (
        await session.scalars(sa.select(Node).filter(Node.name == "two"))
    ).one_or_none()

    print(two.children)

    print(await two.awaitable_attrs.parent)

    await session.close()

```

and in fact it's not even querying the DB because that object "node one" is already in the identity map.

So the next thing going on is as stated elsewhere, "selectinload" and other eager loaders need the "join_depth" setting which indicates "how many levels deep to eager load" and it defaults to zero, that's documented [here](https://docs.sqlalchemy.org/en/20/orm/self_referential.html#self-referential-eager-loading).   so .children loads because you have a join_depth there, but without having it on .parent, .parent won't eager load by default.  so if you want to see a "selectinload" for .parent, you have to put a join depth there also 

```py
    parent: orm.Mapped[Optional["Node"]] = orm.relationship(
        back_populates="children", remote_side=id, lazy="selectin",
        join_depth=1,
    )

```

then the program emits a selectinload for both sides

```
SELECT nodes.id, nodes.name, nodes.parent_id 
FROM nodes 
WHERE nodes.name = ?
2023-07-24 19:36:44,351 INFO sqlalchemy.engine.Engine [generated in 0.00014s] ('two',)
2023-07-24 19:36:44,352 DEBUG sqlalchemy.engine.Engine Col ('id', 'name', 'parent_id')
2023-07-24 19:36:44,352 DEBUG sqlalchemy.engine.Engine Row (2, 'two', 1)
2023-07-24 19:36:44,354 INFO sqlalchemy.engine.Engine SELECT nodes.parent_id AS nodes_parent_id, nodes.id AS nodes_id, nodes.name AS nodes_name 
FROM nodes 
WHERE nodes.parent_id IN (?)
2023-07-24 19:36:44,354 INFO sqlalchemy.engine.Engine [generated in 0.00017s] (2,)
2023-07-24 19:36:44,355 DEBUG sqlalchemy.engine.Engine Col ('nodes_parent_id', 'nodes_id', 'nodes_name')
2023-07-24 19:36:44,356 INFO sqlalchemy.engine.Engine SELECT nodes.id AS nodes_id, nodes.name AS nodes_name, nodes.parent_id AS nodes_parent_id 
FROM nodes 
WHERE nodes.id IN (?)
2023-07-24 19:36:44,356 INFO sqlalchemy.engine.Engine [generated in 0.00015s] (1,)
2023-07-24 19:36:44,357 DEBUG sqlalchemy.engine.Engine Col ('nodes_id', 'nodes_name', 'nodes_parent_id')
2023-07-24 19:36:44,358 DEBUG sqlalchemy.engine.Engine Row (1, 'one', None)
{}
<__main__.Node object at 0x7f68e20aef50>

```   


now.   should join_depth default to "1" in any case with a self-referential eager loader?   maybe.    however, right now it defaults to zero.   the section at https://docs.sqlalchemy.org/en/20/orm/self_referential.html#self-referential-eager-loading definitely needs updating since it does not refer to "selectinload" and also should talk about many-to-ones.  it should also mention new options we have on `selectinload()` for this.

**Reply by ubernostrum** [reply]
Yeah, as noted above I figured out through tinkering with it that `join_depth` had to be set for both "sides" of the relation.

I can put together a PR to update the docs/examples to make that explicit, if you'd like.

**Reply by zzzeek** [reply]
Yes that would be great, thanks very much James !
